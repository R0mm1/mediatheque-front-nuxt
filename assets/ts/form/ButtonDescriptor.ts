import AbstractDescriptor from '@/assets/ts/form/AbstractDescriptor'
import ButtonHrefDescriptor from '~/assets/ts/form/ButtonHrefDescriptor'

export type ButtonStyle = 'normal' | 'negative' | 'icon-round';
export type ButtonRoundedCorner = Boolean | Number; // Number to specify the value to use in CSS, true to use default value, false to deactivate

export default class ButtonDescriptor extends AbstractDescriptor {
  static descriptorType = 'ButtonDescriptor'

  static readonly typeButton = 'button'
  static readonly typeSubmit = 'submit'

  descriptorType = ButtonDescriptor.descriptorType
  type: ('button' | 'submit') = ButtonDescriptor.typeButton
  value: string = ''
  disabled: boolean = false
  customClasses: string[] = []
  style: ButtonStyle = 'normal'
  noDefaultStyle: boolean = false
  href?: ButtonHrefDescriptor = undefined

  constructor (name: string, value?: string, type?: string) {
    super(name)

    if (typeof value !== 'undefined') {
      this.value = value
    }

    if (typeof type !== 'undefined') {
      if (type !== ButtonDescriptor.typeButton && type !== ButtonDescriptor.typeSubmit) {
        throw new Error('Invalid type ' + type + ' provided')
      }
      this.type = type
    }
  }

  addCustomClass (customClass: string) {
    this.customClasses.push(customClass)
    return this
  }

  setValue (value: string) {
    this.value = value
    return this
  }

  setNoDefaultStyle (noDefaultStyle: boolean) {
    this.noDefaultStyle = noDefaultStyle
    return this
  }

  setHref (href: ButtonHrefDescriptor) {
    this.href = href
    return this
  }

  setStyle (style: ButtonStyle) {
    this.style = style
    return this
  }
}
