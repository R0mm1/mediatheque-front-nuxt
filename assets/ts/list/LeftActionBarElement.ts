import AbstractDescriptor from '@/assets/ts/form/AbstractDescriptor'

export default class LeftActionBarElement {
    callback!: Function;
    elementDescriptor!: AbstractDescriptor;

    constructor (callback: Function, elementDescriptor: AbstractDescriptor) {
      this.callback = callback
      this.elementDescriptor = elementDescriptor
    }
}
