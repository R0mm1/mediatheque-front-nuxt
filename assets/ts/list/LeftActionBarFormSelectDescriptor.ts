import SelectDescriptor from '@/assets/ts/form/SelectDescriptor'

export default class LeftActionBarFormSelectDescriptor extends SelectDescriptor {
  static descriptorType = 'LeftActionBarFormSelectDescriptor';

  descriptorType = LeftActionBarFormSelectDescriptor.descriptorType;

  constructor (name: string, options: { [index: string]: string }, label: string = '') {
    super(name, label)

    this.options = options
    this.noLabel = true

    return this
  }
}
