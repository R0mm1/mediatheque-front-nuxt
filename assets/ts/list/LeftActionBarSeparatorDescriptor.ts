import AbstractDescriptor from '~/assets/ts/form/AbstractDescriptor'

export default class LeftActionBarSeparatorDescriptor extends AbstractDescriptor {
  descriptorType = 'LeftActionBarSeparatorDescriptor';
}
