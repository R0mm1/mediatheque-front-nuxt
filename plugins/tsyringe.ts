import { container } from 'tsyringe'
import { NuxtApp } from '@nuxt/types/app'
import Auth from '~/assets/ts/config/public/Auth'
import Api from '~/assets/ts/config/public/Api'
import Tokens from '~/assets/ts/config/Public'

export default ({ app }: {app: NuxtApp}) => {
  container.register<Auth>(Tokens.auth, { useValue: app.$config.auth })
  container.register<Api>(Tokens.api, { useValue: app.$config.api })
  container.register(Tokens.default, { useValue: app.$config.default })
}
