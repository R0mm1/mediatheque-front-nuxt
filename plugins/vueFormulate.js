import Vue from 'vue'

import VueFormulate from '@braid/vue-formulate'
import Chips from '../components/form/elements/formulate/Chips'
import Files from '../components/form/elements/formulate/Files'

Vue.component('Chips', Chips)
Vue.component('Files', Files)

Vue.use(VueFormulate, {
  library: {
    chips: {
      classification: 'chips',
      component: Chips,
      slotProps: {
        component: ['entities', 'entityFields', 'fieldsSeparator', 'formCreationValidationAction', 'formCreationTitle', 'searchFieldPlaceholder', 'searchParam', 'entityURI', 'formCreationSchema', 'editModeOn']
      }
    },
    files: {
      classification: 'files',
      component: Files,
      slotProps: {
        component: ['name', 'maxFiles', 'downloadAction', 'files', 'onFileAdded', 'onFileRemoved']
      }
    }
  }
})
