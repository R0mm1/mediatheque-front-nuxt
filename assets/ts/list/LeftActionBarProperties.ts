import LeftActionBarElement from '@/assets/ts/list/LeftActionBarElement'
import LeftActionBarSeparator from '@/assets/ts/list/LeftActionBarSeparator'

export default class LeftActionBarProperties {
    hasAddButton: boolean;

    customElements: (LeftActionBarElement | LeftActionBarSeparator)[];

    constructor (customElements: (LeftActionBarElement | LeftActionBarSeparator)[] = [], hasAddButton: boolean = true) {
      this.customElements = customElements
      this.hasAddButton = hasAddButton
    }
}
