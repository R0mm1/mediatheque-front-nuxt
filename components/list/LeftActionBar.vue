<template>
  <div id="leftActionBar">
    <div id="labColorBar" />
    <div id="labElements">
      <div
        v-for="(labElement) in labElements"
        :key="labElement.formElementDescriptor.name"
        class="leftActionBarElement"
        :class="getClass(labElement)"
      >
        <div class="leftActionBarElementIcon" :class="labElement.formElementDescriptor.faIcon" />

        <div class="leftActionBarElementText">
          <MedInputButton
            v-if="isButton(labElement)"
            :button-descriptor="labElement.formElementDescriptor"
            @click.native="customFilterChangeHandler(labElement)"
          />

          <MedInputSelect
            v-if="isSelect(labElement)"
            :ref="labElement.formElementDescriptor.name"
            v-model="customFiltersValues[labElement.formElementDescriptor.name]"
            :select-descriptor="labElement.formElementDescriptor"
          />

          <NuxtLink v-if="isLink(labElement)" :to="labElement.formElementDescriptor.to">
            {{ labElement.formElementDescriptor.label }}
          </NuxtLink>

          <div v-if="isSeparator(labElement)" class="leftActionBarSeparatorText">
            {{ labElement.formElementDescriptor.label }}
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import LeftActionBarProperties from '../../assets/ts/list/LeftActionBarProperties'
import ButtonDescriptor from '~/assets/ts/form/ButtonDescriptor'
import LeftActionBarElement from '~/assets/ts/list/LeftActionBarElement'
import listModule from '~/assets/ts/store/ListModule'
import LeftActionBarFormSelectDescriptor from '~/assets/ts/list/LeftActionBarFormSelectDescriptor'
import Filter from '~/assets/ts/list/Filter'
import MedInputSelect from '~/components/form/elements/MedInputSelect.vue'
import MedInputButton from '~/components/form/elements/MedInputButton.vue'
import LeftActionBarLinkDescriptor from '~/assets/ts/list/LeftActionBarLinkDescriptor'

const customFilterChangeHandler = function (element: LeftActionBarElement, value?: any) {
  if (element.type === 'filter') {
    switch (element.formElementDescriptor.descriptorType) {
      case LeftActionBarFormSelectDescriptor.descriptorType:
        listModule.setCustomFilter(new Filter(element.formElementDescriptor.name, value))
        break
      case ButtonDescriptor.descriptorType:
        listModule.setCustomFilter(new Filter(element.formElementDescriptor.name, element.callback()))
        break
    }
  } else {
    element.callback()
  }
}

@Component({
  components: { MedInputSelect, MedInputButton }
})
export default class LeftActionBar extends Vue {
  @Prop(Object) leftActionBarProperties!: LeftActionBarProperties

  get labElements () {
    const elements = []

    if (this.leftActionBarProperties.hasAddButton) {
      elements.push(new LeftActionBarElement(
        'element',
        () => {
          this.$parent.$emit('list-action-add')
          return null
        },
        new ButtonDescriptor('add', 'Ajouter').setFaIcon('fas fa-plus').setNoDefaultStyle(true)
      ))
    }

    elements.push(...this.leftActionBarProperties.customElements)
    return elements
  }

  isButton (element: LeftActionBarElement) {
    return element.formElementDescriptor.descriptorType === 'ButtonDescriptor'
  }

  isSelect (element: LeftActionBarElement) {
    return element.formElementDescriptor.descriptorType === 'LeftActionBarFormSelectDescriptor'
  }

  isSeparator (element: LeftActionBarElement) {
    return element.formElementDescriptor.descriptorType === 'LeftActionBarSeparatorDescriptor'
  }

  isLink (element: LeftActionBarElement) {
    return element.formElementDescriptor.descriptorType === LeftActionBarLinkDescriptor.descriptorType
  }

  getClass (element: LeftActionBarElement) {
    const descriptorBasedClass = element.formElementDescriptor.descriptorType.split('Descriptor').shift()
    if (typeof descriptorBasedClass === 'string') {
      return 'labe-' + descriptorBasedClass[0].toLowerCase() + descriptorBasedClass.substr(1)
    }
    return 'labe-unknown'
  }

  get listCustomFilters () {
    return listModule._customFilters
  }

  /**
   * Return the custom filters, indexed by the name of the filter and wrapped in a proxy.
   * This way, each custom filter can refer to himself in v-model so:
   * - when the value is read, the proxy return the current filter value from the list module, or the default defined in the form descriptor when appropriate
   * - when the value is updated, the proxy will update the filter value in the list module
   */
  get customFiltersValues () {
    const customFilters = this.listCustomFilters

    const handler = {
      get (obj: {[key: string]: LeftActionBarElement}, prop: string) {
        const lfb = obj[prop] ?? undefined
        if (typeof lfb !== 'undefined') {
          if (typeof customFilters[lfb.formElementDescriptor.name] !== 'undefined') {
            return customFilters[lfb.formElementDescriptor.name].value
          }

          if (lfb.formElementDescriptor.descriptorType === 'LeftActionBarFormSelectDescriptor') {
            return (lfb.formElementDescriptor as LeftActionBarFormSelectDescriptor).defaultValue
          }
        }
        return null
      },
      set (obj: {[key: string]: LeftActionBarElement}, prop: string, value: string) {
        const lfb = obj[prop] ?? undefined

        if (typeof lfb !== 'undefined') {
          customFilterChangeHandler(lfb, value)
        }

        return true
      }
    }

    const indexedCustomElements: {[key: string]: LeftActionBarElement} = {}
    this.leftActionBarProperties.customElements.forEach((element) => {
      indexedCustomElements[element.formElementDescriptor.name] = element
    })

    return new Proxy<{[key: string]: LeftActionBarElement}>(indexedCustomElements, handler)
  }

  customFilterChangeHandler (element: LeftActionBarElement, value?: any) {
    customFilterChangeHandler(element, value)
  }
}
</script>

<style scoped lang="scss">
@import "../../assets/scss/breakpoints";

@include phone-portrait {
  #leftActionBar {
    display: none;
  }
}

#labColorBar {
  height: 100%;
  width: 30px;
  position: fixed;
  background-color: #e4dccc;
}

#labElements {
  position: relative;
  overflow: hidden;
  height: 100%;
}
</style>

<style lang="scss">
@import "assets/scss/colors";

.leftActionBarElement {
  line-height: 30px;
  height: 30px;
  cursor: pointer;
  z-index: 1;
  display: flex;

  .leftActionBarElementIcon {
    width: 30px;
    min-width: 30px;
    text-align: center;
    display: inline-block;
    transition: background-color .3s, color .3s;
    font-size: 1rem;
    height: 30px;
    line-height: 30px;
  }

  .leftActionBarElementText {
    transition: background-color .3s;
    width: 100%;
    padding-left: 3px;

    .formulate-input, .formulate-input-wrapper, .formulate-input-element {
      height: 100%;

      button {
        border: 0;
        width: 100%;
        height: 100%;
        text-align: left;
        display: block;
        background-color: transparent;

        i {
          display: none;
        }
      }

      .formulate-input-element--select {
        position: relative;

        &::before {
          content: "";
          width: 0;
          height: 0;
          border-color: black transparent transparent;
          border-style: solid;
          border-width: .3em .3em 0;
          right: 1em;
          position: absolute;
          top: 50%;
          transform: translateY(-50%);
        }

        select {
          -webkit-appearance: none;
          -moz-appearance: none;
          width: 100%;
          border: none;
          height: 100%;
          background: top;
        }
      }

    }

    a{
      font-size: .9rem;
      text-decoration: none;
      color: inherit;
      padding-left: 4px;
    }
  }

  &:not(.labe-leftActionBarSeparator):hover .leftActionBarElementIcon {
    background-color: #bbaf99;
  }

  &:not(.labe-leftActionBarSeparator):hover .leftActionBarElementText {
    background-color: #e4dccc;
  }

  &.labe-leftActionBarSeparator {
    cursor: default;
    overflow: hidden;
    color: $textDisabled;

    .leftActionBarSeparatorIcon {
      background-color: #eeeae1;
    }

    .leftActionBarSeparatorText {
      flex: 1;
      line-height: 30px;
      padding-left: 5px;
    }
  }
}
</style>
