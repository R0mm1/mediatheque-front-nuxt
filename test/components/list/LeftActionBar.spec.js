import 'reflect-metadata'
import { shallowMount } from '@vue/test-utils'
import '~/plugins/vueFormulate'
import LeftActionBar from '~/components/list/LeftActionBar.vue'
import LeftActionBarProperties from '~/assets/ts/list/LeftActionBarProperties'
import LeftActionBarElement from '~/assets/ts/list/LeftActionBarElement'
import LeftActionBarSeparatorDescriptor from '~/assets/ts/list/LeftActionBarSeparatorDescriptor'
import ButtonDescriptor from '~/assets/ts/form/ButtonDescriptor'
import LeftActionBarFormSelectDescriptor from '~/assets/ts/list/LeftActionBarFormSelectDescriptor'

describe('LeftActionBar', () => {
  const wrapper = shallowMount(LeftActionBar, {
    propsData: {
      leftActionBarProperties: new LeftActionBarProperties([
        new LeftActionBarElement(
          'separator',
          () => null,
          new LeftActionBarSeparatorDescriptor('customAdds').setLabel('Ajouter').setFaIcon('fas fa-plus')
        ),
        new LeftActionBarElement(
          'element',
          () => null,
          new ButtonDescriptor('addPaper', 'Livre papier').setFaIcon('fas fa-scroll').setNoDefaultStyle(true)
        ),
        new LeftActionBarElement(
          'element',
          () => null,
          new ButtonDescriptor('addElectronic', 'Epub').setFaIcon('fas fa-tablet-alt').setNoDefaultStyle(true)
        ),
        new LeftActionBarElement(
          'separator',
          () => null,
          new LeftActionBarSeparatorDescriptor('filters').setLabel('Filtres').setFaIcon('fas fa-sliders-h')
        ),
        new LeftActionBarElement(
          'filter',
          () => null,
          new LeftActionBarFormSelectDescriptor('bookType', {
            all: 'Tous',
            paper: 'Papier',
            electronic: 'Epub'
          }).setDefaultValue('all').setFaIcon('fas fa-book').setNoDefaultStyle(true)
        )
      ], true)
    }
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  test('is rendered correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
