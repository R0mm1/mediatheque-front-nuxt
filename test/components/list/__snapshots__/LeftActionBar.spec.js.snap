// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftActionBar is rendered correctly 1`] = `
<div
  id="leftActionBar"
>
  <div
    id="labColorBar"
  />
   
  <div
    id="labElements"
  >
    <div
      class="leftActionBarElement labe-button"
    >
      <div
        class="leftActionBarElementIcon fas fa-plus"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <medinputbutton-stub
          buttondescriptor="[object Object]"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="leftActionBarElement labe-leftActionBarSeparator"
    >
      <div
        class="leftActionBarElementIcon fas fa-plus"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <!---->
         
        <!---->
         
        <div
          class="leftActionBarSeparatorText"
        >
          
          Ajouter
        
        </div>
      </div>
    </div>
    <div
      class="leftActionBarElement labe-button"
    >
      <div
        class="leftActionBarElementIcon fas fa-scroll"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <medinputbutton-stub
          buttondescriptor="[object Object]"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="leftActionBarElement labe-button"
    >
      <div
        class="leftActionBarElementIcon fas fa-tablet-alt"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <medinputbutton-stub
          buttondescriptor="[object Object]"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="leftActionBarElement labe-leftActionBarSeparator"
    >
      <div
        class="leftActionBarElementIcon fas fa-sliders-h"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <!---->
         
        <!---->
         
        <div
          class="leftActionBarSeparatorText"
        >
          
          Filtres
        
        </div>
      </div>
    </div>
    <div
      class="leftActionBarElement labe-leftActionBarFormSelect"
    >
      <div
        class="leftActionBarElementIcon fas fa-book"
      />
       
      <div
        class="leftActionBarElementText"
      >
        <!---->
         
        <medinputselect-stub
          selectdescriptor="[object Object]"
        />
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;
